File: src/api/schemas/drafts.py
"""Draft API schemas — list, detail, approve, reject, reassign."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from src.api.schemas.emails import ClassificationSummary


class DraftListItem(BaseModel):
    """Single draft in a paginated list."""

    id: uuid.UUID
    email_id: uuid.UUID
    email_subject: str
    email_sender: str
    status: str
    reviewer_id: uuid.UUID | None
    created_at: datetime


class EmailForDraftReview(BaseModel):
    """Email context for side-by-side draft review. No body_plain (PII)."""

    id: uuid.UUID
    subject: str
    sender_email: str
    sender_name: str | None
    snippet: str | None
    received_at: datetime
    classification: ClassificationSummary | None = None


class DraftDetailResponse(BaseModel):
    """Full draft detail for review."""

    id: uuid.UUID
    content: str
    status: str
    reviewer_id: uuid.UUID | None
    reviewed_at: datetime | None
    pushed_to_provider: bool
    email: EmailForDraftReview
    created_at: datetime
    updated_at: datetime


class DraftApproveRequest(BaseModel):
    """Request body for POST /drafts/{id}/approve."""

    push_to_gmail: bool = True


class DraftApproveResponse(BaseModel):
    """Response after approving a draft."""

    draft_id: uuid.UUID
    approved: bool
    gmail_draft_id: str | None = None
    approved_at: datetime
    note: str | None = None


class DraftRejectRequest(BaseModel):
    """Request body for POST /drafts/{id}/reject."""

    reason: str = Field(min_length=1)


class DraftReassignRequest(BaseModel):
    """Request body for POST /drafts/{id}/reassign."""

    reviewer_id: uuid.UUID

File: frontend/src/api/drafts.ts
// src/api/drafts.ts
// Typed API functions for draft endpoints — types from generated schema
import apiClient from "./client";
import type {
  PaginationParams,
  PaginatedResponse,
  DraftListItem,
  DraftDetailResponse,
  DraftApproveRequest,
  DraftApproveResponse,
  DraftRejectRequest,
  DraftReassignRequest,
} from "@/types/generated/api";

/** Query params accepted by GET /drafts */
type DraftListParams = PaginationParams & {
  status?: "pending" | "approved" | "rejected";
};

/**
 * GET /drafts — paginated draft list with optional status filter.
 */
export async function fetchDrafts(
  params: DraftListParams,
): Promise<PaginatedResponse<DraftListItem>> {
  const { data } = await apiClient.get<PaginatedResponse<DraftListItem>>("/drafts", {
    params,
  });
  return data;
}

/**
 * GET /drafts/{draftId} — full draft detail with email context for side-by-side review.
 */
export async function fetchDraftDetail(draftId: string): Promise<DraftDetailResponse> {
  const { data } = await apiClient.get<DraftDetailResponse>(`/drafts/${draftId}`);
  return data;
}

/**
 * POST /drafts/{draftId}/approve — approve a draft and optionally push to Gmail.
 */
export async function approveDraft(
  draftId: string,
  body?: DraftApproveRequest,
): Promise<DraftApproveResponse> {
  const { data } = await apiClient.post<DraftApproveResponse>(
    `/drafts/${draftId}/approve`,
    body ?? {},
  );
  return data;
}

/**
 * POST /drafts/{draftId}/reject — reject a draft with a required reason.
 * Returns HTTP 204 No Content — resolves to void.
 */
export async function rejectDraft(draftId: string, body: DraftRejectRequest): Promise<void> {
  await apiClient.post(`/drafts/${draftId}/reject`, body);
}

/**
 * POST /drafts/{draftId}/reassign — reassign draft to a different reviewer (admin only).
 */
export async function reassignDraft(
  draftId: string,
  body: DraftReassignRequest,
): Promise<DraftDetailResponse> {
  const { data } = await apiClient.post<DraftDetailResponse>(
    `/drafts/${draftId}/reassign`,
    body,
  );
  return data;
}
